assets/js/src/index.js

```diff
@@ -1,6 +1,24 @@
-// Initialize tracking for classic WooCommerce pages
-import { trackClassicPages } from './integrations/classic';
-window.wcgai.trackClassicPages = trackClassicPages;
+import { createTracker } from './tracker';
+import { classicTracking } from './integrations/classic';
+import { blocksTracking } from './integrations/blocks';
 
-// Initialize tracking for Block based WooCommerce pages
-import './integrations/blocks';
+// Wait for DOMContentLoaded to make sure event data is in place.
+if ( document.readyState === 'loading' ) {
```

Comment:
I took this construct from [MDN example](https://developer.mozilla.org/en-US/docs/Web/API/Document/DOMContentLoaded_event#checking_whether_loading_is_already_complete) 

----

> If the inline script for data (window.ga4w) is also optimized and delayed in loading then wouldn't the chance be bigger that we throw the error during this `interactive` stage.

Do you mean the case when inline data is deferred to the `interactive` phase, while our JS may run before?

I made a few synthetic test cases to achieve such effect:
- wrapping the inline data inside `setTimeout` https://stackblitz.com/edit/inline-script-in-interactive-document-state?file=index.html
- wrapping inline data inside inline **async module** https://stackblitz.com/edit/inline-script-in-interactive-document-state-rcxnmd?file=index.html

But do you think we should support that?
1. If somebody is messing with the scripts by changing them from blocking to asynchronous, there are bigger problems to expect
3. I didn't test it yet, but I wonder if running on "load" may be too late if we want to catch some events, before page redirects, etc.


---

On the other hand, if we're going to prepare ourselves to be run async, we could as well deliberately set up ourselves using async modules. I didn't test the new API to this extend yet. Especially since there were some discussions regarding the lack of inline module support in WP.  But then we could do something like:

```html
<!-- WP-build import map -->
<script type="importmap">
  {
    "imports": {
      "woocommerce-google-analytics-integration": "path/to/main.js",
    }
  }
</script>
<script id="woocommerce-google-analytics-integration" type="module" src="path/to/main.js" async></script>
<script type="module" async>
import initializeTracking from 'woocommerce-google-analytics-integration';

const ga4w = { /* inlined JSON */}
initializeTracking( ga4w );
</script>
```


But still, what "optimizations" for such scripts we should anticipate, I don't know.

Reply:
I was referring to some of the more aggressive options in plugins like Autoptimize:
![image](https://github.com/woocommerce/woocommerce-google-analytics-integration/assets/11388669/4223e8f0-f163-4780-898d-ea7dbf6d9c24)

I haven't actually tested all those scenarios, and regardless of how it's aggregated, I'd expect the load order to remain the same within the aggregated script. Although looking at the forum there was plenty cases where the load order was changed as well [as plugins](https://wordpress.org/plugins/async-javascript/) that purely allow you to set `async`, which means we can no longer guarantee load order.

> But do you think we should support that?
> 1. If somebody is messing with the scripts by changing them from blocking to asynchronous, there are bigger problems to expect
> 2. I didn't test it yet, but I wonder if running on "load" may be too late if we want to catch some events, before page redirects, etc.

I wouldn't expect us to be able to support every scenario, but as long as we are compatible with a handful of main optimizers and there are options that can be set to exclude particular scripts then I'd find that sufficient.

> may be too late if we want to catch some events, before page redirects

As far as I've seen the Google gtag script has a delay (at least in my cases a couple seconds) after load so page redirects that happen upon load would almost always be too late to catch.

I like the async module solution. It looks like a stable solution, although I haven't worked with that before, so unable to determine if that would have any secondary effects.

Reply:
Thanks for clarifications, I'll do more experiments with Autooptimize.

> I like the async module solution. It looks like a stable solution, although I haven't worked with that before, so unable to determine if that would have any secondary effects.

I did some experiments, but i think it's too early to do so:
- Script Modules API is only available in WP 6.5 (not yet released)
- it does not support inline modules - we could not pass the data that way
- modules cannot have script dependencies - we could not add the classic `google-tag-manager` script as a dependency

Reply:
Sounds like modules are still a little limiting.
We need to always register a script with the `google-tag-manager` ID, as otherwise WC core will load it's fallback. So we'd have to keep that in mind for resolving dependencies. But as you mentioned it's not ready yet, so at most I'd add the information we currently know to an enhancement issue so it's easy to track.

Reply:
I made an issue for ESModules approach https://github.com/woocommerce/woocommerce-google-analytics-integration/issues/411

and tweaked the wait for data flow in https://github.com/woocommerce/woocommerce-google-analytics-integration/pull/398/commits/dacaabcf29aff66f310fd71f90cd5f335a7d5285. So it no longer throws, but use the data once it comes. Which make GA4W kinda work with Autoptimize "Aggregate inline JS". But that option puts dynamic data into cached file, which is a separate problem.

Reply:
Thanks I like that solution.